### PR TITLE
fix overflow bug in `first(::OffsetArray, ::Integer)`

### LINKED ISSUE #45842: `first` sometimes returns incorrect results when used with nonstandard indexing
The function `first(itr, n)` is defined to get the first `n` elements of the collection `itr`, or fewer elements if `itr` is not long enough.

It behaves properly with plain arrays:

```jl
julia> a = [2, 3, 4];

julia> first(a, typemax(Int))
3-element Vector{Int64}:
 2
 3
 4
```

However, it returns the wrong answer with offset arrays:

```jl
julia> using OffsetArrays

julia> b = OffsetArray(a, 2:4);

julia> first(b, typemax(Int))
Int64[]
```

The culprit is this “faster method for vectors”:

https://github.com/JuliaLang/julia/blob/bc9fdcef490f32c48e92fe93c373b2aa7076e55b/base/abstractarray.jl#L467-L471

The expression `begin + n - 1` overflows and becomes negative, causing the array to be incorrectly indexed with an empty index range.

```
julia> versioninfo()
Julia Version 1.7.3
Commit 742b9abb4d (2022-05-06 12:58 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin21.4.0)
  CPU: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)
  ```

### PR DESCRIPTION
**Pull request by oscardssmith**

### CODE REVIEW

**Comment by oscardssmith** [comment]
yes. Would the correct definition in general be `A[eachindex(A)[1:n]]`?

**Comment by KristofferC** [comment]
More explicit PR titles would be nice.

File: base/abstractarray.jl [hunk 1]
```diff
@@ -467,7 +467,10 @@ first(itr, n::Integer) = collect(Iterators.take(itr, n))
 # Faster method for vectors
 function first(v::AbstractVector, n::Integer)
     n < 0 && throw(ArgumentError("Number of elements must be nonnegative"))
-    @inbounds v[begin:min(begin + n - 1, end)]
+    stop = firstindex(v) + n - 1,
+    stop < firstindex(v) && (stop = lastindex(v))
```
Comment (mikmoore) [inline base/abstractarray.jl:471]: This line does not work correctly for `n=0`.

**Comment by giordano** [comment]
I hope it goes without saying that any bugfix should come with tests.  As a side note, this method is already tested for offset arrays at https://github.com/JuliaLang/julia/blob/84bf42a9a57f42df141e488444e728c9510bf359/test/offsetarray.jl#L631-L644 (but not with pathological second arguments, which is what the issue is about)

File: base/abstractarray.jl [hunk 2]
```diff
@@ -467,7 +467,10 @@ first(itr, n::Integer) = collect(Iterators.take(itr, n))
...
+    stop =  min(stop, lastindex(v))
+    @inbounds v[begin:stop]
```
Comment (jishnub) [inline base/abstractarray.jl:473]: Is this equivalent to `v[range(begin, length=min(n, length(v)))]`?

**Comment by jishnub** [comment]
Presumably, `last` needs to be fixed as well:
```julia
julia> b = OffsetArray(a, typemin(Int));

julia> last(a, 100)
3-element Vector{Int64}:
 2
 3
 4

julia> last(b, 100)
Int64[]
```

**Comment by jishnub** [comment]
@oscardssmith gentle bump. In case you don't have time to work on this, would it be ok if I continue it?

**Comment by oscardssmith** [comment]
oh I totally forgot about this. if you want to finish it, go ahead.

**Pull request closed without merging**